File: packages/core-py/agentid/rotation.py
from __future__ import annotations

from datetime import datetime, timezone

from nacl.exceptions import BadSignatureError
from nacl.signing import SigningKey, VerifyKey

from agentid.identity import get_agent_id
from agentid.types import Keypair, RotationStatement
from agentid.utils import decode_base58, decode_base64, encode_base64


def create_rotation(old_keypair: Keypair, new_public_key: bytes) -> RotationStatement:
    """The old key signs a delegation statement to the new key."""
    old_key_id = get_agent_id(old_keypair.public_key)
    new_key_id = get_agent_id(new_public_key)
    rotated_at = datetime.now(timezone.utc).isoformat()

    continuity_proof = f"rotate:{old_key_id}:{new_key_id}:{rotated_at}"
    signing_key = SigningKey(old_keypair.private_key)
    signed = signing_key.sign(continuity_proof.encode("utf-8"))

    return RotationStatement(
        type="rotation",
        old_key_id=old_key_id,
        new_key_id=new_key_id,
        rotated_at=rotated_at,
        continuity_proof=continuity_proof,
        signature_by_old_key=encode_base64(signed.signature),
    )


def verify_rotation(statement: RotationStatement) -> tuple[bool, list[str]]:
    errors: list[str] = []

    expected_proof = f"rotate:{statement.old_key_id}:{statement.new_key_id}:{statement.rotated_at}"
    if statement.continuity_proof != expected_proof:
        errors.append("Continuity proof content does not match statement fields")

    old_public_key = _extract_public_key(statement.old_key_id)
    if old_public_key is None:
        return False, ["Cannot extract public key from old_key_id"]

    proof_bytes = statement.continuity_proof.encode("utf-8")
    signature_bytes = decode_base64(statement.signature_by_old_key)

    try:
        verify_key = VerifyKey(old_public_key)
        verify_key.verify(proof_bytes, signature_bytes)
    except (BadSignatureError, Exception):
        errors.append("Signature by old key is invalid")

    return len(errors) == 0, errors


def _extract_public_key(key_id: str) -> bytes | None:
    prefix = "aid_ed25519_"
    if not key_id.startswith(prefix):
        return None
    try:
        return decode_base58(key_id[len(prefix):])
    except Exception:
        return None

File: packages/core-ts/src/rotation.ts
import * as ed25519 from '@noble/ed25519';
import type { Keypair, RotationStatement } from './types.js';
import { getAgentId } from './identity.js';
import { encodeBase64, decodeBase64, decodeBase58 } from './utils.js';

/**
 * Create a signed rotation statement proving key continuity.
 * The old key signs a statement delegating trust to the new key.
 */
export async function createRotation(
  oldKeypair: Keypair,
  newPublicKey: Uint8Array
): Promise<RotationStatement> {
  const oldKeyId = getAgentId(oldKeypair.publicKey);
  const newKeyId = getAgentId(newPublicKey);
  const rotatedAt = new Date().toISOString();

  const continuityProof = `rotate:${oldKeyId}:${newKeyId}:${rotatedAt}`;
  const proofBytes = new TextEncoder().encode(continuityProof);
  const signatureBytes = await ed25519.signAsync(proofBytes, oldKeypair.privateKey);

  return {
    type: 'rotation',
    old_key_id: oldKeyId,
    new_key_id: newKeyId,
    rotated_at: rotatedAt,
    continuity_proof: continuityProof,
    signature_by_old_key: encodeBase64(signatureBytes),
  };
}

/**
 * Verify a rotation statement: check that the old key actually signed the
 * continuity proof delegating to the new key.
 */
export async function verifyRotation(
  statement: RotationStatement
): Promise<{ valid: boolean; errors: string[] }> {
  const errors: string[] = [];

  const expectedProof = `rotate:${statement.old_key_id}:${statement.new_key_id}:${statement.rotated_at}`;
  if (statement.continuity_proof !== expectedProof) {
    errors.push('Continuity proof content does not match statement fields');
  }

  const oldPublicKey = extractPublicKey(statement.old_key_id);
  if (!oldPublicKey) {
    return { valid: false, errors: ['Cannot extract public key from old_key_id'] };
  }

  const proofBytes = new TextEncoder().encode(statement.continuity_proof);
  const signatureBytes = decodeBase64(statement.signature_by_old_key);

  try {
    const signatureValid = await ed25519.verifyAsync(signatureBytes, proofBytes, oldPublicKey);
    if (!signatureValid) {
      errors.push('Signature by old key is invalid');
    }
  } catch {
    errors.push('Signature verification threw an error');
  }

  return { valid: errors.length === 0, errors };
}

function extractPublicKey(keyId: string): Uint8Array | null {
  const prefix = 'aid_ed25519_';
  if (!keyId.startsWith(prefix)) return null;
  try {
    return decodeBase58(keyId.slice(prefix.length));
  } catch {
    return null;
  }
}
